MapboxCoreNavigation/RouteController.swift

```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
         }
 
         // Only check every so often for a faster route.
-        guard location.timestamp.timeIntervalSince(lastLocationDate) >= RouteControllerProactiveReroutingInterval else {
+        guard location.timestamp.timeIntervalSince(lastLocationDate) >= routeControllerProactiveReroutingInterval else {
             return
         }
+        
         let durationRemaining = routeProgress.durationRemaining
         
         isFindingFasterRoute = true
+        
+        print("[RouteController] Checking for faster/updated route...")
 
-        getDirections(from: location, along: routeProgress) { [weak self] (route, error) in
-            guard let strongSelf = self else {
-                return
-            }
+        getDirections(from: location, along: routeProgress) { [weak self] mostSimilarRoute, routes, error in
+            guard let self = self else { return }
             
             // Every request should reset the lastLocationDate, else we spam the server by calling this method every location update.
             // If the call fails, tough luck buddy! Then wait until the next interval before retrying
-            strongSelf.lastLocationDate = nil
+            self.lastLocationDate = nil
             
             // Also only do one 'findFasterRoute' call per time
-            strongSelf.isFindingFasterRoute = false
-
-            guard let route = route else {
+            self.isFindingFasterRoute = false
+            
+                
+            guard let route = mostSimilarRoute, let routes = routes else {
                 return
             }
-
-            guard let firstLeg = route.legs.first, let firstStep = firstLeg.steps.first else {
-                return
+            
+            self.applyNewRerouteIfNeeded(mostSimilarRoute: route, allRoutes: routes, currentUpcomingManeuver: currentUpcomingManeuver, durationRemaining: durationRemaining)
+        }
+    }
+    
+    func applyNewRerouteIfNeeded(mostSimilarRoute: Route, allRoutes: [Route], currentUpcomingManeuver: RouteStep, durationRemaining: TimeInterval) {
+        guard let firstLeg = mostSimilarRoute.legs.first, let firstStep = firstLeg.steps.first else {
+            return
+        }
+        
+        // Current and First step of old and new route should be significant enough of a maneuver before applying a faster route, so we don't apply the route just before a maneuver will occur
+        let significantFirstStep = firstStep.expectedTravelTime >= RouteControllerMediumAlertInterval && routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval
+        
+        // Current maneuver should correspond to the next maneuver in the new route
+        let sameUpcomingManeuver = currentUpcomingManeuver == firstLeg.steps[1]
```

Comment:
Are you sure there are always at least 2 steps in a leg? Otherwise this array will go out of bounds and crash.

Reply:
If you are sure, please comment :)

Reply:
This behaviour isn't changed from the existing rerouting logic and I think routes always have a 'start' maneuver at the beginning. Not sure though, I can make this more safe